internal/urlutil/url.go

```diff
@@ -83,8 +83,8 @@ func GetAbsoluteURL(r *http.Request) *url.URL {
 //
 // For standard HTTP (80)/HTTPS (443) ports, it returns `example.com` and `example.com:<port>`.
 // Otherwise, return the URL.Host value.
-func GetDomainsForURL(u *url.URL) []string {
-	if IsTCP(u) {
+func GetDomainsForURL(u url.URL) []string {
```

Comment:
Why not a pointer here (and elsewhere)?

Reply:
- I fixed a subtle pointer-related bug that occurred when multiple upstreams were present in the config, and that started changing pointers to non-pointers backwards - there were more in previous commits 
- use of pointers style is officially discouraged by the Go team https://github.com/golang/go/wiki/CodeReviewComments#pass-values
- indeed if you read thru i.e. K8S or any other protobuf centric code, there would be tons of pointers passed around because they explicitly prevent that, however that does not make that idiomatic code style. 
- pointers are potential source of faults - in principle you need check they're not nil
- i.e. I missed a check upon a condition in `options.Validate()` and that resulted in SIGSEGV in so many functions all over the place that I gave up fixing them. For a server software it's preferable not to be in a crash loop, IMHO. 
- in such functions they rarely provide any saving memory or GC wise; they're allocated on stack and cost nothing

Reply:
Thanks for the context and I take your point.

I think `url.URL` is a small enough struct to not really matter, and to your point, avoids many of the foot-guns related to passing pointers. 

But, my reading of the "pass-values" code review comments is not that the generic use of pointers is officially discouraged, but rather 

> "Don't pass pointers as function arguments just to save a few bytes."

With the caveat that 

> This advice does not apply to large structs, or even small structs that might grow.

To be fair, [url.URL](https://golang.org/pkg/net/url/#URL) is not large, but it's not tiny ( like the given example `*string`) either. 

I've gone back and forth about whether passing `url.URL` as a pointer made sense, and ended up defaulting to what I saw most often in the standard library which is to use a pointer. The example that jumps immediately to mind is [http.Request](https://golang.org/pkg/net/http/#Request). 

 I agree with all your other points.